### ISSUE TITLE
rethinkDNS using a lot of data in firewall only mode

### ISSUE THREAD

**Issue by x-tk** [celzero/rethink-app#766]
Good afternoon, after some commentary on the .53n release thread on reddit ( https://www.reddit.com/r/rethinkdns/comments/10b0k9v/comment/j6x9klx/?context=3 ), I am posting this issue here.

I got a new pixel 7 pro for black friday and so got a clean slate on this phone.  Moved over stuff manually to the new phone including a new rethinkDNS install.   I am using the firewall only mode (no dns queries) because I run internal dns and don't want to open up external dns and deal with switching between services when I go between the house.

I am seeing very high data usage for the app in the Android data usage section.  From dec 8 to jan 7, my data is showing rethinkDNS using 165mb.  From jan 8 to feb 7, my data is showing rethinkDNS is using 774mb.  

I have Google FI so I pay for my data, so this is concerning,  Wanted to file an issue and work with you guys to find out what is going on and see if there is a bug or something causing this.  Let me know if there's anything you need me to provide.

Thanks!

**Comment by ignoramous** [comment]
700MB+ in a month? Wow. This is concerning. Would you know if other app's data use is accounted on Rethink (which is a local VPN)?

**Comment by x-tk** [comment]
Yeah, my reaction as well haha.  872mb now for jan 8 to feb 7.  the next closest are android os (105mb), instagram (42mb), weather (25mb), yahoo fantasy (25mb), etc etc.  other apps are showing on the list as well.  I do have a personal vpn at the house but have not setup this phone on it yet, so it can't be counting that.  I  have used proton vpn a couple of times in the month but pretty sparingly and not since that last usage I provided you the other day where I went from 774 to 872mb used now this cycle.

**Comment by hussainmohd-a** [comment]
@x-tk : Can you please provide the screenshot of the **_Storage_** info of Rethink app?
You can find it in _Rethink -> About -> App Info -> Storage & cache_

**Comment by x-tk** [comment]
Yes, attached
![Screenshot_20230206-120713](https://user-images.githubusercontent.com/17145865/217038180-b93e4891-f111-4dac-ba2f-07fe898f8052.png)

**Comment by x-tk** [comment]
Nothing?

**Comment by x-tk** [comment]
That is encouraging at least.  The problem is that I have Google Fi and it's billing me based on usage and it shows this is usage counted towards my bill?  So what am I supposed to do?

**Comment by ignoramous** [comment]
> The problem is that I have Google Fi (which bills on data used) and it's billing me based on usage and it shows this is usage counted towards my bill? So what am I supposed to do?

Rethink isn't, to our knowledge, generating that much amount of traffic on its own. Rethink however is responsible for forwarding traffic from installed apps to its proper destination (when it doesn't forward that traffic is how an app or IP is firewalled). Its our theory that Android wrongly attributes that traffic as coming from Rethink itself.

You can try the *DNS-only* mode (which disables Firewall and hence the traffic forwarding part) if these traffic stats have you skeptical about Rethink.

This issue is open as we want to thoroughly investigate once app version `v055` (due April/May) is out the door.

**Comment by x-tk** [comment]
Sounds good, SUPER appreciate you guys looking into it!  Yeah the thing is, I don't want to use the DNS mode.  I host my own DNS at my house and have to toggle it on/off leaving the house if I don't want to expose my DNS servers publicly.. So that is the issue, I am really only using it for firewall.  I used to use netguard and did not have this problem but the application is a bit clunkier interface wise.  Love the app in general - just hoping to get this resolved!

**Comment by x-tk** [comment]
Good morning -- with the new release, is there any update in the code that would check this?  Just checking to see if I should take another look at it again. Thank you.

**Comment by ignoramous** [comment]
Hi, no new updates. We're fully focused on #52 (for `v055` since it is a feature that should have launched 20 months ago) and immediately after that, this bug is next.

**Comment by x-tk** [comment]
Makes sense, thank you.

**Comment by x-tk** [comment]
Glad I am not the only one :sweat_smile:

**Comment by ignoramous** [comment]
A user (not using Pixel but using Awaken A11) wrote to us,

> As rethink requires vpn connection to run so due to "Allow hotspot clients to use VPN" option the system recognised it as VPN but allowed my PC to use Rethink's tunnel on my phone :) so basically that's not Rethink's data consumption but the PC's using the Hotspot.

**Comment by x-tk** [comment]
With google fi, my monthly payments line up with the carrier usage on the phone which includes the amount that Rethink DNS claims to be using within the OS.  This is anywhere from 600-875mb the last few months on the carrier usage tracking.  Other apps are showing usage as well so it's hard to determine if that is just from times where rethink was turned off or if there is truly a usage issue.   This was not an issue with netguard as app mobile usage always reported correctly.

**Comment by ignoramous** [comment]
>  This is anywhere from 600-875mb the last few months on the carrier usage tracking. 

An astonishing number, really. Makes me want to cry.

We do *pad* DNS questions to [128b](https://github.com/celzero/firestack/blob/4576e99a96bacb3b349e8c73a63e3699f135e3ed/intra/doh/padding.go#L32) in size, but even at 100k queries/mo, the DNS requests alone shouldn't exceed 12MB/mo egress, but you're not even using DNS!

>  This was not an issue with netguard as app mobile usage always reported correctly.

Btw, NetGuard and Rethink have very different architectures. Rethink converts incoming packets into TCP or UDP sockets and then proxies them out (similar to what proxifiers like SingBox + Matsuri / SagerNet do):

```
#send
App ---[send]--> OS ---[write]--> TUN <--[read]--- Rethink ---[multiplex/firewall]--> Sockets (TCP/UDP) ---[duplex/proxy]--> Socket (TCP/UDP) ---[to]--> Destination

#its dual: recv
App <--[recv]--- OS <--[read]--- TUN <--[write]--- Rethink <--[multiplex/firewall]--- Sockets (TCP/UDP) <--[duplex/proxy]--- Socket (TCP/UDP) <--[from]--- Destination
```

NetGuard may be doing something else, I haven't checked.

Will keep looking for answers.

**Comment by x-tk** [comment]
Thanks for continuing to look into this, just curious if there is a new status yet?  Thanks.

**Comment by ignoramous** [comment]
Yes; the thing we consistently see in `/proc/<pid>/net/tcp[6]` and `/proc/<pid>/net/udp[6]` is connections made by Google Play services / Google Play framework attributed to Rethink. Otherwise, there's just one other connection attributed to Rethink's own process open, which is to an upstream DoH server.

DoH is on the heavier side data-wise, so assuming 300K * 2KB DNS responses per month, the total usage comes around 600MB/mo.

Other than that, we suspect [any dependency](https://github.com/celzero/firestack/blob/n2/go.mod) we use might come with a built-in botnet...? The rate of data-use we see is around ~1MB/hour or so.
